Split Trace into sub-namespaces

beniamin opened the pull request:
I wanted to know your opinion about organizing Trace namespace into multiple sub-namespaces. For me, it's simpler to understand and follow the code, especially when I needed to change same thing in multiple classes of the same type(exporters, samplers, processors).

I want to know your opinion. If you are ok with that approach, then I will update the opened PRs to, or I will wait for those to be merged and make the change then.

morrisonlevi commented:
As far as I'm concerned, this is a decrease in quality. Here are two reasons:

1. When something like a `Sampler` is in its own namespace it tends to import a bunch of things like `Context`, `Span`, `Attributes`, `Links` etc. Generally, more boilerplate is worse and this causes more boilerplate (boilerplate is code that is written that doesn't actually _do_ anything, like importing things with `use`).
2. This is a lot of namespace segments and adds redundancy e.g. `OpenTelemetry\Sdk\Trace\Sampler\Sampler`.

With that said, I'm not going to force my opinion on everyone. I just don't see how this adds value (and in my opinion it makes it worse).

-----

Now, the individual exporters are vendor-specific and it makes sense to split them off. In fact, may make sense to eventually split them into their own repositories; this is a clear sign they should have their own namespace. Maybe `OpenTelemetry\Contrib\Zipkin` with `SpanConverter` and `Exporter` in there?

beniamin commented:
1. I can't see where the quality drops. I think the benefit is that it increases the code readability, and you can follow the classes more naturally. It is really confusing to find all Samplers among all other Trace related classes on the same directory. I can't see the increase in `use` statements because the SDK classes depend on the API interfaces, not on the implementations.

2. I am wondering now, why the interfaces like `Exporter`, `Sampler`, `SpanProcessor` are still in the SDK and why these are not part of the API?

I like the idea of having different repositories for exporters, definitely, this will improve the performance of the SDK and will reduce the dependencies, like Guzzle. Moreover, we can go further and have different repositories for each exporter and anyone can choose the exporter to use by simply using one repository or another.

beniamin commented:
I made some changes to my initial proposal, based on @morrisonlevi suggestions.
1. I created a new 'area' for `contrib` and moved the exporters there and also I moved their tests.
2. I moved the Interfaces outside sub-namespaces in order to avoid redundancy like `OpenTelemetry\Sdk\Trace\Sampler\Sampler`.

Do you think is better now?

bobstrecansky commented on contrib/Tests/Unit/JaegerExporterTest.php:
```diff
@@ -2,10 +2,10 @@
 declare(strict_types=1);
 
-namespace OpenTelemetry\Tests\Unit\Exporter;
+namespace OpenTelemetry\Contrig\Tests\Unit;
```
I think this should be `Contrib`, not `Contrig`

beniamin replied:
fixed

bobstrecansky commented on contrib/Tests/Unit/ZipkinExporterTest.php:
```diff
@@ -2,10 +2,10 @@
 declare(strict_types=1);
 
-namespace OpenTelemetry\Tests\Unit\Exporter;
+namespace OpenTelemetry\Contrig\Tests\Unit;
```
I think this should be `Contrib`, not `Contrig`

beniamin replied:
fixed

bobstrecansky commented on sdk/Tests/Unit/Trace/TracingTest.php:
```diff
@@ -2,7 +2,7 @@
 
```
We were working on leaving the tests in the `tests` directory - was this an inadvertent move?  I spoke to this more in our gitter channel.

beniamin replied:
done. I moved all tests inside the same root directory

bobstrecansky commented on .phan/config.php:
```diff
@@ -316,7 +316,7 @@
     // The number of processes to fork off during the analysis
     // phase.
-    'processes' => 1,
+    'processes' => 2,
```
Was this changed to improve testing speed?

beniamin replied:
it was just a leftover, I tested to see if I get some speed, but I didn't saw any improvement.

bobstrecansky commented on contrib/Tests/Unit/ZipkinSpanConverterTest.php:
```diff
@@ -2,12 +2,12 @@
 declare(strict_types=1);
 
-namespace unit\Exporter;
+namespace OpenTelemetry\Contrig\Tests\Unit;
```
I think this should be `Contrib`, not `Contrig`

beniamin replied:
done

bobstrecansky commented on examples/AlwaysOnTraceExample.php:
```diff
@@ -3,13 +3,13 @@
 require __DIR__ . '/../vendor/autoload.php';
 
-use OpenTelemetry\Sdk\Trace\AlwaysOnSampler;
+use OpenTelemetry\Contrib\Zipkin\Exporter;
```
When I ran the examples locally with this branch, I got an error (included below) - I think we need to maybe `use` this as another name, because it's conflicting with `sdk/Trace/Exporter.php` - Maybe we should import it `as ZipkinExporter` or something like that?

```
bentest ? make examples                                                                                                                                                                                                                                                                                                                                                   2 ↵
docker-compose up -d
Starting opentelemetry-php_php_1 ...
opentelemetry-php_zipkin_1 is up-to-date
Starting opentelemetry-php_php_1 ... done
docker-compose run --rm php php ./examples/AlwaysOnTraceExample.php

Fatal error: Uncaught Error: Class 'OpenTelemetry\Contrib\Zipkin\Exporter' not found in /usr/src/myapp/examples/AlwaysOnTraceExample.php on line 24

Error: Class 'OpenTelemetry\Contrib\Zipkin\Exporter' not found in /usr/src/myapp/examples/AlwaysOnTraceExample.php on line 24

Call Stack:
    0.0011     405464   1. {main}() /usr/src/myapp/examples/AlwaysOnTraceExample.php:0

make: *** [examples] Error 255```

beniamin replied:
I imported Zipking/Exporter to ZipkinExporter. I couldn't reproduce this error before, but can you try it now and see if it works?

For me it works ok. I also tried to remove all library directory and reinstall it, but I couldn't reproduce the error.

```bash
Beniamins-MacBook➜  opentelemetry : organized-trace-sdk ✔ : ᐅ  make examples                                                                                                                                                      (emag-prod-admin@content-prod/content)
docker-compose up -d
Creating opentelemetry_php_1    ... done
Creating opentelemetry_zipkin_1 ... done
Creating opentelemetry_jaeger_1 ... done
docker-compose run --rm php php ./examples/AlwaysOnTraceExample.php
Starting AlwaysOnTraceExample
Trace with id 00609987d713d0f9e434a7fbf10513b5 started 
AlwaysOnTraceExample complete!  See the results at http://localhost:9411/
docker-compose run --rm php php ./examples/AlwaysOffTraceExample.php

Sampling is not enabled
docker-compose run --rm php php ./examples/JaegerExporterExample.php
Starting JaegerExporterExample
Trace with id 3e8da8026b581b89f508589ef8c97f73 started [{"id":"338a016600bcb55e","traceId":"3e8da8026b581b89f508589ef8c97f73","parentId":"6204daafba5ee7fe","localEndpoint":{"serviceName":"jaegerExporterExample"},"name":"session.generate.span1590676668","timestamp":1590676668901991,"duration":114021120,"tags":{"remote_ip":"1.2.3.4","country":"USA"},"annotations":[{"timestamp":1590676668901991,"value":"found_login0"},{"timestamp":1590676668901991,"value":"generated_session"}]}][{"id":"33b0c7074625d1db","traceId":"3e8da8026b581b89f508589ef8c97f73","parentId":"6204daafba5ee7fe","localEndpoint":{"serviceName":"jaegerExporterExample"},"name":"session.generate.span1590676669","timestamp":1590676669934031,"duration":928000,"tags":{"remote_ip":"1.2.3.4","country":"USA"},"annotations":[{"timestamp":1590676669934031,"value":"found_login1"},{"timestamp":1590676669934031,"value":"generated_session"}]}][{"id":"690cde2c2ac341b6","traceId":"3e8da8026b581b89f508589ef8c97f73","parentId":"6204daafba5ee7fe","localEndpoint":{"serviceName":"jaegerExporterExample"},"name":"session.generate.span1590676669","timestamp":1590676669969406,"duration":760064,"tags":{"remote_ip":"1.2.3.4","country":"USA"},"annotations":[{"timestamp":1590676669969406,"value":"found_login2"},{"timestamp":1590676669969406,"value":"generated_session"}]}][{"id":"39e0f62e1516c569","traceId":"3e8da8026b581b89f508589ef8c97f73","parentId":"6204daafba5ee7fe","localEndpoint":{"serviceName":"jaegerExporterExample"},"name":"session.generate.span1590676670","timestamp":1590676670026732,"duration":581120,"tags":{"remote_ip":"1.2.3.4","country":"USA"},"annotations":[{"timestamp":1590676670026732,"value":"found_login3"},{"timestamp":1590676670026732,"value":"generated_session"}]}][{"id":"57441e6a26d91f61","traceId":"3e8da8026b581b89f508589ef8c97f73","parentId":"6204daafba5ee7fe","localEndpoint":{"serviceName":"jaegerExporterExample"},"name":"session.generate.span1590676670","timestamp":1590676670069244,"duration":4576000,"tags":{"remote_ip":"1.2.3.4","country":"USA"},"annotations":[{"timestamp":1590676670069244,"value":"found_login4"},{"timestamp":1590676670069244,"value":"generated_session"}]}]
JaegerExporterExample complete!  See the results at http://localhost:16686/
Beniamins-MacBook➜  opentelemetry : organized-trace-sdk ✔ : ᐅ                                              
```

bobstrecansky replied:
Yup; this worked for me now too.

beniamin commented:
Is there anything I should do?

bobstrecansky commented:
@beniamin - I'm okay with this, but I was waiting for @morrisonlevi to comment, as I know he had opinions on this.

morrisonlevi commented:
I've been busy lately and haven't been able to follow what the Resource is about. A cursory glance in the spec says it's part of the SDK -- if so it probably should be in the SDK namespace, not its own namespace, shouldn't it? Again, have been busy so sorry if this was already discussed elsewhere.

beniamin commented:
> I've been busy lately and haven't been able to follow what the Resource is about. A cursory glance in the spec says it's part of the SDK -- if so it probably should be in the SDK namespace, not its own namespace, shouldn't it? Again, have been busy so sorry if this was already discussed elsewhere.

From the specs I understand that Resource is an Entity used by both Trace SDK and Metrics SDK and should be kept separately. So I am ok with keeping Resource in a different namespace.

From the specs:
> When used with distributed tracing, a resource can be associated with the TracerProvider when it is created. That association cannot be changed later. When associated with a TracerProvider, all Spans produced by any Tracer from the provider MUST be associated with this Resource.
> 
> Analogous to distributed tracing, when used with metrics, a resource can be associated with a MeterProvider. When associated with a MeterProvider, all metrics produced by any Meter from the provider will be associated with this Resource.

beniamin commented:
> Just read through all the comments and changes here. My opinion is that its a net-neutral change. A little bit of organization, but it doesn't feel like namespace-bloat or namespaces with just one class. The namespaces aren't large, but maybe that's ok?
> 
> * I like the `contrib` split, that did confuse me a bit when I was first diving into the repo structure. The Samplers and SpanProcessors getting split off also feels pretty natural since there are 3-4 of each and grouping them makes sense.
> * As for `resource`, the OT example project layout also includes it at the top-level of the `sdk` structure: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-layout.md#sdk-package so it makes sense to do likewise.
> * I think `SpanMultiProcessor` needs to get moved into `sdk/Trace/SpanProcessor/`... what about the `SpanProcessor` interface, should that stay in the SDK or should it get moved to the API?

For the last bullet, I am happy to move SpanMultiProcessor in the `sdk/Trace/SpanProcessor/`. For the `SpanProcessor` Interface I asked on specifications channel on Gitter and the answer was:
> the goal of the API is to provide an interfaces for objects are being used to instrument libraries and things like that
> that classes you mentioned are used to create your application tracer
which is irrelevant for API

beniamin commented on examples/JaegerExporterExample.php:
```diff
@@ -20,16 +20,15 @@
 );
 
-$exporter = new JaegerExporter(
+$exporter = new JaegerExpoter(
```
fixed

beniamin commented on examples/JaegerExporterExample.php:
```diff
@@ -3,11 +3,11 @@
 require __DIR__ . '/../vendor/autoload.php';
 
-use OpenTelemetry\Sdk\Trace\AlwaysOnSampler;
+use OpenTelemetry\Contrib\Jaeger\Exporter as JaegerExpoter;
```
fixed
